messages/compiler/cpp/test_cppgen.py

```diff
@@ -0,0 +1,294 @@
+    # What we do because of this is just generate a single KV PAIR for now. Generating a single
+    # pair just means using double brackets for map_start and map_end. We can think of more
+    # sophisticated strategies later.
+    def map_start(self):
+        self.instance += "{{"
+
+    def map_key_end(self):
+        self.instance += ","
+
+    def map_end(self):
+        self.instance += "}}"
+
+    def optional_start(self):
+        self.instance += "{"
+
+    def optional_end(self):
+        self.instance += "}"
+
+    def oneof(self, msgs):
+        name = random.choice(list(msgs.keys()))
+        variable_instance = random.choice(self.existing_instances[name])
+        self.instance += type_instance_from_variable_instance(
+            variable_instance)
+
+
+def testSerializationStr(msg_name):
+    """
+    Create a function that roundtrip serializes and deserializes all instances of a given message type.
+    """
+    s = "void {}() {{\n".format(test_name(msg_name))
+    for i in range(0, MAX_SIZE):
+        instance = instance_name(msg_name, i)
+        s += """
+  {{
+    std::vector<uint8_t> output;
+    serialize(output, {});
+    {} {}_computed;
+    deserialize(output, {}_computed);
+    assert({} == {}_computed);
+  }}
+""".format(instance, msg_name, instance, instance, instance, instance)
+    s += "}\n"
+    return s
+
+
+def file_header(namespace):
+    return """/***************************************
+ Autogenerated by test_cppgen.py. Do not modify.
+***************************************/
+
+#include <cassert>
+#include "example.h"
+
+namespace {} {{
+
+""".format(namespace)
+
+
+def file_trailer(namespace, ast):
+    s = """
+}} // namespace {}
+
+int main() {{
+""".format(namespace)
+    for msg in ast.msgs:
+        s += "  {}::{}();\n".format(namespace, test_name(msg.name))
+    s += "}"
+    return s
+
+
+def generate_code_and_tests(ast):
+    """ Walk concord message format(CMF) AST and generate C++ code and C++ tests"""
+    namespace = "cmf::test"
+    print("Generating C++ Message structs and serialization code")
+    code = cppgen.translate(ast, namespace)
+    test_code = file_header(namespace)
+    print("Generating C++ message instances and serialization tests")
+    visitor = InstanceVisitor()
+    # We generate `max_size` msg instances for tests
+    for i in range(0, MAX_SIZE):
+        visitor.size = i
+        walker = Walker(ast, visitor)
+        walker.walk()
+    for msg_name, instances in visitor.existing_instances.items():
+        for instance in instances:
+            test_code += instance + "\n\n"
+        test_code += testSerializationStr(msg_name)
+    return code, test_code + file_trailer(namespace, ast)
+
+
+def compile_tests():
+    print("Compiling tests with g++")
+    os.system(
+        f"g++ -std=c++17 -g -o {OUTPUT_DIR}/test_serialization {OUTPUT_DIR}/test_serialization.cpp"
+    )
+
+
+def run_tests():
+    print("Running tests")
+    if os.system("./test_serialization") == 0:
```

Comment:
This needs to be `f"./{OUTPUT_DIR}/test_serialization"`

Reply:
Ahh, silly last minute changes to use an output directory. Good catch!

Reply:
fixed.